### PR TITLE
S expr

### LINKED ISSUE #1132: move s expression parsing to main library
Move the s expression parsing code from the Python library to the main Arbor library.

The s-expression parser would not be part of the public API: the public API would use strings. Consequently, any code in the Python wrapper that uses the s-expression feature would also have to be ported, which means that the following changes need to be made
  * move s-expression definition header file to back end library
  * move s-expression implementation and lexer source to back end library
  * move label expression parsing code to back end library with a stringified public interface

The existing unit tests can also be moved over.

### LINKED ISSUE #1139: rename region/locset expressions to use `-` instead of `_`

### PR DESCRIPTION
**Pull request by bcumming**
Move s expression parsing of strings, used for parsing region and locset expressions, from the Python wrapper to the main library.

CMake
* Add `ARBDEV_COLOR` option that forces gcc and clang to always output color output.

Move label parsing code from Python to arbor C++ library
* Replace `hopefully` type implemented in python headers with an `arb::util::expected`.
* Update `locset` and `region` constructors that take string arguments to
      parse strings as s-expressions or "quoted" labels.
* updated the input stream modifier used to parse asc files to use a lookup table
      for substitutions, renamed it transmogrifier.                                                             

Tests:
* moved s expression parsing tests into `unit` and extended them.

General label concepts
* Allow arbitrary strings in labels.
* Add `parse_region_expression` and `parse_locset_expression` functions
      alongside the existing `parse_label_expression` function for use when
      a region or locset is expected. These calls will promote a quoted string
      `"label"` to `(region "label")` or `(locset "label")` respectively.
* Add string literals for labels so that the C++ interface can use 
      `"soma"_lab` instead of awkward escaping `"\"soma\""`.

Python:
* Remove old label parsing code.
* Simplify Python wrapper code.

General:
* replaced some `arb::util::any` and `arb::util::variant` instances with their `std::` version.

Fixes #1139
Fixes #1132

### CODE REVIEW

**Comment by bcumming** [comment]
bors try

**Review by noraabiakar — COMMENTED** [review]
Looks good! I have just a couple of small questions. Also the examples need updating.

File: test/unit/test_morph_stitch.cpp [hunk 1]
```diff
@@ -60,9 +62,9 @@ TEST(morph, stitch_two) {
         EXPECT_EQ(p2, seg1.prox);
         EXPECT_EQ(p3, seg1.dist);
 
-        mprovider p(m, sm.labels("stitch:"));
-        EXPECT_TRUE(region_eq(p, "stitch:0", reg::segment(0)));
-        EXPECT_TRUE(region_eq(p, "stitch:1", reg::segment(1)));
+        mprovider p(m, sm.labels("stitch-"));
+        EXPECT_TRUE(region_eq(p, "stitch-0"_lab, reg::segment(0)));
```
[the remark this answers is not in the document]
Comment (bcumming) [reply test/unit/test_morph_stitch.cpp:66]: There was (it isn't a valid character in an s-expression symbol), however that has been relaxed. I will change back.

File: arbor/s_expr.hpp [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+struct token {
+    src_location loc;
+    tok kind;
+    std::string spelling;
+};
+
+std::ostream& operator<<(std::ostream&, const token&);
+
+std::vector<token> tokenize(const char* line);
```
[the remark this answers is not in the document]
Comment (bcumming) [reply arbor/s_expr.hpp:173]: It is not. I remove it.

**Comment by bcumming** [comment]
bors try

**Review by halfflat — APPROVED** [review]
Approving with the new tests.

**Pull request merged**